File: chapter_3/Deck/Dealer.java
package chapter_3.Deck;

public class Dealer {
    private Deck deck;

    public Dealer(Deck d) {
        deck = d;
    }

    public void shuffle() {
        int num_cards = deck.size();

        for (int i = 0; i < num_cards; i++) {
            int index = (int)(Math.random() * num_cards);
            Card card_i = (Card)deck.get(i);
            Card card_index =  (Card)deck.get(index);
            deck.replace(i, card_index);
            deck.replace(index, card_i);
        }
    }

    public Card dealCard() {
        if (deck.size() > 0) {
            return deck.removeFromFront();
        }

        return null;
    }
}

File: chapter_3/DeckPy/dealer.py
from copy import copy
from deck import Deck
from random import shuffle


class Dealer:
    def __init__(self):
        self.deck = Deck()

    def shuffle(self):
        _deck = copy(self.deck)
        shuffle(_deck)
        self.deck = _deck

    def deal_card(self):
        if self.deck:
            return self.deck.remove_from_front()

    def __str__(self):
        return str(self.deck)

    def __repr__(self):
        return str(self.deck)


class DishonestDealer(Dealer):
    def shuffle(self):
        pass

    def deal_card(self):
        if self.deck:
            return self.deck.get(0)